### PR TITLE
Error handling for `accelerator="mps"` and `ddp` strategy pairing

### LINKED ISSUE #16148: Error handling for `accelerator="mps"` and `strategy="ddp"`
### Bug description

The MPS backend does not support `torch.distributed`. We should fail early in the AcceleratorConnector and produce a user friendly error message.

### How to reproduce the bug

```python
import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule, Trainer


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


def run():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    model = BoringModel()
    trainer = Trainer(
        accelerator="mps",
        devices=1,
        strategy="ddp"
    )
    trainer.fit(model, train_dataloaders=train_data)


if __name__ == "__main__":
    run()
```

### Error messages and logs

```
Traceback (most recent call last):
  File "/Users/adrian/repositories/lightning/examples/pl_bug_report/bug_report_model.py", line 49, in <module>
    run()
  File "/Users/adrian/repositories/lightning/examples/pl_bug_report/bug_report_model.py", line 45, in run
    trainer.fit(model, train_dataloaders=train_data)
  File "/Users/adrian/repositories/lightning/src/pytorch_lightning/trainer/trainer.py", line 598, in fit
    call._call_and_handle_interrupt(
  File "/Users/adrian/repositories/lightning/src/pytorch_lightning/trainer/call.py", line 36, in _call_and_handle_interrupt
    return trainer.strategy.launcher.launch(trainer_fn, *args, trainer=trainer, **kwargs)
  File "/Users/adrian/repositories/lightning/src/pytorch_lightning/strategies/launchers/subprocess_script.py", line 88, in launch
    return function(*args, **kwargs)
  File "/Users/adrian/repositories/lightning/src/pytorch_lightning/trainer/trainer.py", line 640, in _fit_impl
    self._run(model, ckpt_path=self.ckpt_path)
  File "/Users/adrian/repositories/lightning/src/pytorch_lightning/trainer/trainer.py", line 1074, in _run
    self.strategy.setup(self)
  File "/Users/adrian/repositories/lightning/src/pytorch_lightning/strategies/ddp.py", line 159, in setup
    self._share_information_to_prevent_deadlock()
  File "/Users/adrian/repositories/lightning/src/pytorch_lightning/strategies/ddp.py", line 405, in _share_information_to_prevent_deadlock
    self._share_pids()
  File "/Users/adrian/repositories/lightning/src/pytorch_lightning/strategies/ddp.py", line 423, in _share_pids
    pids = self.all_gather(torch.tensor(os.getpid(), device=self.root_device))
  File "/Users/adrian/repositories/lightning/src/pytorch_lightning/strategies/parallel.py", line 90, in all_gather
    return _all_gather_ddp_if_available(tensor, group=group, sync_grads=sync_grads)
  File "/Users/adrian/repositories/lightning/src/lightning_lite/utilities/distributed.py", line 202, in _all_gather_ddp_if_available
    return _AllGather.apply(tensor, group)
  File "/Users/adrian/repositories/lightning/src/lightning_lite/utilities/distributed.py", line 170, in forward
    torch.distributed.all_gather(gathered_tensor, tensor, group=group)
  File "/Users/adrian/miniconda3/envs/lightning/lib/python3.10/site-packages/torch/distributed/distributed_c10d.py", line 2277, in all_gather
    work = group.allgather([tensor_list], [tensor])
RuntimeError: ProcessGroupGloo::allgather: unsupported device type mps
```

This applies for the entire DDPStrategy family (ddp, ddp_fork, ddp_spawn, etc.)

### Environment

<details>
  <summary>Current environment</summary>

```
#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow): Trainer, LightningLite
#- PyTorch Lightning Version (e.g., 1.5.0): 1.9.0dev
#- Lightning App Version (e.g., 0.5.2): x
#- PyTorch Version (e.g., 1.10): 1.13
#- Python version (e.g., 3.9): 3.10
#- OS (e.g., Linux): MacOS
#- CUDA/cuDNN version: x
#- GPU models and configuration: x
#- How you installed Lightning(`conda`, `pip`, source): pip 
#- Running environment of LightningApp (e.g. local, cloud): x
```

</details>

### More info

_No response_

cc @borda @justusschock

### PR DESCRIPTION
**Pull request by shenoynikhil**
## What does this PR do?
Fixes #16148 

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 1]
```diff
@@ -283,6 +283,30 @@ def _check_config_and_set_final_flags(
...
+                            DDPSpawnShardedStrategy,
+                            DDPShardedStrategy,
+                            DDPFullyShardedNativeStrategy,
+                            DDPFullyShardedStrategy,
+                            DDPSpawnStrategy,
+                        ),
+                    )
+                )
+            )
+        ):
```
Comment (Borda) [inline src/pytorch_lightning/trainer/connectors/accelerator_connector.py:307]:
```suggestion
        ddp_classes = (
                            DDPStrategy,
                            DDPSpawnShardedStrategy,
                            DDPShardedStrategy,
                            DDPFullyShardedNativeStrategy,
                            DDPFullyShardedStrategy,
                            DDPSpawnStrategy,
                        )
        is_ddp_str = isinstance(strategy, str) and "ddp" in strategy
        if (
            accelerator is not None
            and strategy is not None
            and (isinstance(accelerator, str) and accelerator == "mps")
            and (is_ddp_str or (isinstance(strategy, ddp_classes))
            )
        ):
```

**Review by Borda — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 2]
```diff
@@ -283,6 +283,25 @@ def _check_config_and_set_final_flags(
...
+                            DDPFullyShardedStrategy,
+                            DDPSpawnStrategy,
+                        )
+        is_ddp_str = isinstance(strategy, str) and "ddp" in strategy
+        if (
+            accelerator is not None
+            and strategy is not None
+            and (isinstance(accelerator, str) and accelerator == "mps")
+            and (is_ddp_str or (isinstance(strategy, ddp_classes))
+            )
```
Comment (Borda) [inline src/pytorch_lightning/trainer/connectors/accelerator_connector.py:301]:
```suggestion
            and (is_ddp_str or isinstance(strategy, ddp_classes)
            )
```

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 3]
```diff
@@ -283,6 +283,24 @@ def _check_config_and_set_final_flags(
...
+            DDPFullyShardedNativeStrategy,
+            DDPFullyShardedStrategy,
+            DDPSpawnStrategy,
+        )
+        is_ddp_str = isinstance(strategy, str) and "ddp" in strategy
+        if (
+            accelerator is not None
+            and strategy is not None
+            and (isinstance(accelerator, str) and accelerator == "mps")
+            and (is_ddp_str or isinstance(strategy, ddp_classes))
```
Comment (justusschock) [inline src/pytorch_lightning/trainer/connectors/accelerator_connector.py:300]:
```suggestion
            and (is_ddp_str or isinstance(strategy, ParallelStrategy))
```
this should cover all kinds of parallelism

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 4]
```diff
@@ -283,6 +283,24 @@ def _check_config_and_set_final_flags(
...
 
+        # mps accelerator incompatible with ddp-family
+        ddp_classes = (
+            DDPStrategy,
+            DDPSpawnShardedStrategy,
+            DDPShardedStrategy,
+            DDPFullyShardedNativeStrategy,
+            DDPFullyShardedStrategy,
+            DDPSpawnStrategy,
+        )
```
Comment (justusschock) [inline src/pytorch_lightning/trainer/connectors/accelerator_connector.py:294]:
```suggestion
```

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 5]
```diff
@@ -283,6 +283,16 @@ def _check_config_and_set_final_flags(
                 f" Available names are: {', '.join(self._accelerator_types)}."
             )
 
+        # mps accelerator incompatible with ddp-family
+        is_ddp_str = isinstance(strategy, str) and "ddp" in strategy
+        if (
+            accelerator is not None
+            and strategy is not None
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/accelerator_connector.py:290]: These 2 are not required, no? The `isinstance(None, str)` will shortcut

```suggestion
```

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 6]
```diff
@@ -283,6 +283,16 @@ def _check_config_and_set_final_flags(
...
+            and (isinstance(accelerator, str) and accelerator == "mps")
+            and (is_ddp_str or isinstance(strategy, ParallelStrategy))
+        ):
+            raise MisconfigurationException("With `accelerator=mps`, strategies from DDP Family are not compatible")
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/accelerator_connector.py:294]:
```suggestion
            raise ValueError("With `accelerator=mps`, strategies from DDP Family are not compatible")
```

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 7]
```diff
@@ -283,6 +283,16 @@ def _check_config_and_set_final_flags(
                 f" Available names are: {', '.join(self._accelerator_types)}."
             )
 
+        # mps accelerator incompatible with ddp-family
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/accelerator_connector.py:286]: @awaelchli does Lite need this change too?

Comment (awaelchli) [reply src/pytorch_lightning/trainer/connectors/accelerator_connector.py:286]: Yes exactly. @shenoynikhil The equivalent change needs to be added to src/lightning_lite/connector.py
Feel free to add it this PR, or you can also leave it for a separate PR.

File: tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py [hunk 8]
```diff
@@ -369,6 +369,33 @@ def test_exception_invalid_strategy():
...
+        ("ddp_sharded_spawn", DDPSpawnShardedStrategy),
+    ),
+)
+@RunIf(mps=True)
+def test_exception_invalid_ddp_strategy_with_mps(strategy, strategy_class):
+    # check with str
+    with pytest.raises(
+        MisconfigurationException, match=r"With `accelerator=mps`, strategies from DDP Family are not compatible"
+    ):
+        Trainer(accelerator="mps", strategy=strategy)
```
Comment (awaelchli) [inline tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py:390]: Edge case alert! We also need to handle `accelerator="auto"` :))

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 6 above)
Comment (awaelchli) [inline src/pytorch_lightning/trainer/connectors/accelerator_connector.py:294]:
```suggestion
            raise MisconfigurationException("With `accelerator=mps`, strategies from the DDP Family are not compatible.")
```

**Review by awaelchli — COMMENTED** [review]
Looks great so far!

File: tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py (hunk 8 above)
Comment (justusschock) [reply tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py:390]: and `accelerator="gpu"` but at some point they will be set to "mps" , iirc

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 5 above)
Comment (shenoynikhil) [reply src/pytorch_lightning/trainer/connectors/accelerator_connector.py]: Incorporated this.

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 6 above)
Comment (shenoynikhil) [reply src/pytorch_lightning/trainer/connectors/accelerator_connector.py]: Incorporated this.

File: tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py (hunk 8 above)
Comment (shenoynikhil) [reply tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py:388]: I have added a case for the `accelerator="auto"` condition. For the edge case @justusschock mentioned, there are other test cases that fail when I add another check. I have commented it out for now,
```
        # MPSAccelerator incompatible with ddp-family
        # check if strategy is string and contains 'ddp' in it
        is_ddp_str = isinstance(strategy, str) and "ddp" in strategy
        # MPS accelerator could be instantiated in 3 ways,
        # 1. accelerator = "mps"
        # 2. accelerator in ("auto", None) and MPSAccelerator.is_available()
        # 3. accelerator = "gpu" and MPSAccelerator.is_available()
        is_mps_accelerator = (
            (isinstance(accelerator, str) and accelerator == "mps")
            or (
                ((isinstance(accelerator, str) and accelerator == "auto") or accelerator is None)
                and MPSAccelerator.is_available()
            )
            # or (isinstance(accelerator, str) and accelerator == "gpu" and MPSAccelerator.is_available())
        )
        if is_mps_accelerator and (is_ddp_str or isinstance(strategy, ParallelStrategy)):
            raise ValueError("With MPSAccelerator, strategies from the DDP family are not compatible.")
```
Can you give some feedback as to why you think the commented-out section causes some test cases to fail?

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 7 above)
Comment (shenoynikhil) [reply src/pytorch_lightning/trainer/connectors/accelerator_connector.py]: I'll add it in this PR once the conversations are resolved.

**Comment by Borda** [comment]
seems like last two conversations to finish this PR :)

File: tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py (hunk 8 above)
Comment (awaelchli) [reply tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py]: @shenoynikhil Finally I got to look at this again. I handled the additional edge case and new ones too, and simplified the boolean logic as much as possible. I updated tests that needed updating. 

Notably: Some tests that only instantiated Trainer but not run fit would now error with our newly added validation, when run on an M1 mac. I handled all these cases by either simulating no mps devices in the test, or skipping the test on mps hardware completely.

**Review by awaelchli — APPROVED** [review]

File: tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py (hunk 8 above)
Comment (shenoynikhil) [reply tests/tests_pytorch/trainer/connectors/test_accelerator_connector.py]: > Some tests that only instantiated Trainer but not run fit would now error with our newly added validation, when run on an M1 mac.

That makes sense. I noticed these when I was writing tests. Wasn't sure if I should modify the others. Skipping them on mps hardware makes sense.

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 7 above)
Comment (awaelchli) [reply src/pytorch_lightning/trainer/connectors/accelerator_connector.py]: The tests seem to pass. I'm ok merging this as is and adding the corresponding changes for lightning_fabric/connector.py in a separate PR.

Comment (shenoynikhil) [reply src/pytorch_lightning/trainer/connectors/accelerator_connector.py]: Yes, I can create the other PR based once this merged.

**Comment by awaelchli** [comment]
Requesting re-review from either @carmocca or @justusschock given latest changes.

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 9]
```diff
@@ -284,6 +285,22 @@ def _check_config_and_set_final_flags(
...
             )
 
+        # MPS accelerator is incompatible with DDP family of strategies. It supports single-device operation only.
+        is_ddp_str = isinstance(strategy, str) and "ddp" in strategy
+        is_dp_str = isinstance(strategy, str) and "dp" in strategy
+        is_parallel_strategy = isinstance(strategy, ParallelStrategy) or is_ddp_str or is_dp_str
+        is_mps_accelerator = MPSAccelerator.is_available() and (
+            accelerator in ("mps", "auto", "gpu", None)
+            or isinstance(accelerator, MPSAccelerator)
+            or isinstance(self._strategy_flag, Strategy)
```
Comment (justusschock) [inline src/pytorch_lightning/trainer/connectors/accelerator_connector.py:295]: why this check? wouldn't this be fine if it is a custom single-device strategy or if you really want parallelism and handle that in a custom strategy via CPU?

**Review by justusschock — APPROVED** [review]

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py (hunk 9 above)
Comment (awaelchli) [reply src/pytorch_lightning/trainer/connectors/accelerator_connector.py:295]: I thought of it as an edge case, but we can also lift this restriction.

Comment (Borda) [reply src/pytorch_lightning/trainer/connectors/accelerator_connector.py:295]: here or another PR? :)

File: src/pytorch_lightning/trainer/connectors/accelerator_connector.py [hunk 10]
```diff
@@ -284,6 +285,19 @@ def _check_config_and_set_final_flags(
                 f" Available names are: {', '.join(self._accelerator_types)}."
             )
 
+        # MPS accelerator is incompatible with DDP family of strategies. It supports single-device operation only.
+        is_ddp_str = isinstance(strategy, str) and "ddp" in strategy
+        is_dp_str = isinstance(strategy, str) and "dp" in strategy
+        is_parallel_strategy = isinstance(strategy, ParallelStrategy) or is_ddp_str or is_dp_str
```
Comment (carmocca) [inline src/pytorch_lightning/trainer/connectors/accelerator_connector.py:291]: Should this check also 
```suggestion
        is_parallel_strategy = isinstance(strategy, ParallelStrategy) or is_ddp_str or is_dp_str or is_deepspeed_str
```

**Review by carmocca — APPROVED** [review]
I assume this will be done to Fabric separately
